src/nublado/startup/services/preparer.py

```diff
@@ -0,0 +1,1193 @@
+"""RSP Lab preparer."""
+
+import asyncio
+import configparser
+import contextlib
+import datetime
+import hashlib
+import json
+import os
+import shutil
+import time
+from pathlib import Path
+from textwrap import dedent
+from typing import Any
+from urllib.parse import parse_qsl, urlparse
+
+import structlog
+import yaml
+from rubin.repertoire import DiscoveryClient, RepertoireError
+from safir.logging import LogLevel, configure_logging
+
+from ..constants import (
+    APP_NAME,
+    ETC_PATH,
+    LAB_STATIC_CMD_ARGS,
+    MAX_NUMBER_OUTPUTS,
+    PREVIOUS_LOGGING_CHECKSUMS,
+    RUNTIME_ENVIRONMENT_VARIABLES,
+    STARTUP_PATH,
+)
+from ..exceptions import RSPErrorCode, RSPStartupError
+from ..storage.command import Command
+from ..utils import (
+    get_access_token,
+    get_digest,
+    get_jupyterlab_config_dir,
+    get_runtime_mounts_dir,
+)
+
+__all__ = ["Preparer"]
+
+
+class Preparer:
+    """Class to prepare for JupyterLab launch by modifying the filesystem
+    and writing out a file representing the runtime environment, to be
+    used by the launcher.
+
+    The launcher resides in lsst.rsp.startup; the user Lab should not have
+    a copy of the nublado class at all.
+    All the launcher does is to read a list of environment variables that
+    the preparer writes, sets those, and then starts the JupyterLab process.
+
+    This setup is very Rubin-specific and opinionated, and will
+    likely not work for anyone else's science platform.
+
+    If that's you, use this for inspiration, but don't expect this to
+    work out of the box.
+    """
+
+    def __init__(self) -> None:
+        # We start with a copy of our own environment.
+        # Only a subset of this will be forwarded into the launch
+        # environment; most of it will be used for internal housekeeping
+        # during the setup process.
+        self._env = os.environ.copy()
+        self._debug = bool(self._env.get("DEBUG", ""))
+        loglevel = LogLevel.DEBUG if self._debug else LogLevel.INFO
+        configure_logging(name=APP_NAME, log_level=loglevel)
+        self._logger = structlog.get_logger(APP_NAME)
+        self._broken = False
+        for req_env in ("JUPYTERHUB_BASE_URL", "NUBLADO_HOME"):
+            if req_env not in self._env:
+                exc = RSPStartupError("EBADENV", None, req_env)
```

Comment:
Do we want to pass in `RSPErrorCode.EBADENV` instead of the string here?

Reply:
Whoops, yes we do.